### ISSUE TITLE
[Tracking]: Watermarking hardening - compatibility tests and quality/performance regression monitoring

### ISSUE THREAD

**Issue by tomasruizt** [vllm-project/vllm#56105]
### 🚀 The feature, motivation and pitch

Follow-up to [the watermarking RFC #53916](https://github.com/vllm-project/vllm/issues/53916) and [the implementation PR #54053](https://github.com/vllm-project/vllm/pull/54053). This issue tracks automated compatibility tests and recurring quality/performance regression checks for watermarking in CI or nightly runs.

The RFC defines the design and reports initial evaluations. This issue tracks turning that validation into ongoing regression coverage as watermarking evolves. We should coordinate compatibility work with the implementation PR's existing tests and unsupported-feature assertions.

## Items:

* Feature compatibility: We want to exercise basic watermarking and detection functionality at a high API level, i.e. from a client perspective: Watermarked and non-watermarked generations are correctly identified as such. These tests should exercise a variety of feature combinations. E.g. speculative decoding, tool calling, structured output, parallelism (e.g. basic TP2 to begin with). The goal is to quickly surface failures when combining features, and provide an easy entrypoint for devs to fix them. Ideally these tests run as quickly as possible, so they might check only a small number of requests.

* Quality Regressions: Watermarking has some theoretical guarantees about distribution preservation, but we must check in practice to prevent regressions. These quality evals are heavier / slower, because they might run an entire dataset with potentially long generations to get statistically valid results. We could use GSM8K, or use a less saturated benchmark, as suggested by @simon-veitner-redhat. The simplest version could be to run the eval once on the watermarked model and compare performance to a reference value with some tolerance (which can be generated once offline). 
    * A more comprehensive approach is to run the entire eval multiple times (one time per key, perhaps 20 keys) to prove robustness over keys. There are statistical tests designed to surface model degradations, e.g.: https://arxiv.org/abs/2602.10144.
    * Note: The [watermarking RFC](https://github.com/vllm-project/vllm/issues/53916) showed a change in generation length, which is unexpected. We hope this disappears as we increase the number of keys n. However, @TQCB found some infinite generation cases in GSM8K, which might be related. If we use an LLM-as-a-judge to evaluate answer quality automatically, infinite generations should be flagged as failures.
    * Note: temperature should be not 0.0, since this deactivates watermarking.

* Speed Regressions: Here we need to benchmark that watermarking stays within a reasonable speed overhead in e2e inference, e.g. 1%. This benchmark could simulate a more realistic inference setup, e.g. use quantization, SD integration (once available), parallelism, and mid or large batch sizes.

* Diversity & Detectability: Any further ideas to monitor this? @TQCB 

## Related work:

* [#54143: Backend-stable stateless RNG for Gumbel sampling](https://github.com/vllm-project/vllm/issues/54143)
* [#52461: Async scheduling and stateful logits processors](https://github.com/vllm-project/vllm/issues/52461)

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by simon-veitner-redhat** [comment]
note that long generation is properly not caused by number of keys but by fact that for equal context gen becomes deterministic. this can be avoided by only once putting the watermark and afterwards sampling "normal".

**Comment by TQCB** [comment]
Diversity:
- unique responses over repeated sampling
- self bleu metrics or similar (wouldn't rely on these but there will be some signal)
- n-gram repetition metrics

We can also measure degenerate responses with changes in response length distribution, ngram repetition metrics, and things like compressing the text + measuring compression ratio. I expect this would be enough signal for most cases and cheaper than running a judge.

**Comment by TQCB** [comment]
For quality evaluation, I've had good results on BeyondAIME. It's quite sensitive to degradation and not particularly noisy (much less than AIME), maybe a good candidate eval.

**Comment by simon-veitner-redhat** [comment]
i like above ideas @TQCB. properly much of the really bad stuff that can happen can be detected with very simply text stats on a simple corpus!

**Comment by simon-veitner-redhat** [comment]
i expect for instance that the worrisome cases will all have degenerate plot if you slide with a 4-window over the whole output and count how much each window (taking order into account) occurs compared to normal generation.

properly we can think of some other heuristics like this and just use them to measure quality.

**Comment by SyaOtiLan** [comment]
Hi @tomasruizt @TQCB, I'd like to help with the speed-regression coverage in this tracker. I saw that #56309 covers the basic end-to-end and GSM8K quality tests, so I'd focus on a reusable benchmark comparing throughput and latency with watermarking on and off.

Is anyone already working on this part? If not, would a focused benchmark for periodic performance checks be useful, and is there an existing benchmark suite you'd prefer it to integrate with?

**Comment by simon-veitner-redhat** [comment]
other benchmarks are much better suited to determine detection ability than GSM8K, see https://github.com/vllm-project/vllm/pull/56122#issuecomment-5640058700

**Comment by tomasruizt** [comment]
HI @SyaOtiLan! We don't have a speed-regression coverage yet. I was thinking about the best way to implement this. In terms of the harness, the perf-eval harness integrates several benchmarks & evals in a single suite (https://github.com/vllm-project/perf-eval/tree/main). Examples are here: https://github.com/vllm-project/perf-eval/blob/main/workloads

While I love this unified infra, it doesn't yet expose assertions on the metrics, so that we could write in the yaml something like this: 
```yaml
...
vllm_bench:
  configs:
    - name: 1k-in-256-out
      ...

      assertions:
        mean_tpot_ms:
          max: 2.5
```

This would allow us to define an absolute target metric to reach, like 2.5ms TPOT, which could be derived from running the baseline offline on the same hardware. However, this requires extending perf-eval with assertions, and its own issue (RFC) probably. I think researching wether this extension has been discussed, fleshing it out and discussing the implementation with others would be valuable!

**Comment by SyaOtiLan** [comment]
Thanks, @tomasruizt — I’ll check for existing discussions around metric assertions in perf-eval and look at how they could fit into the current result-processing flow. I’ll start with the absolute thresholds you suggested, keeping watermarking on/off comparisons as a separate consideration, and share the findings here.

**Comment by tomasruizt** [comment]
@simon-veitner-redhat @TQCB To swap out GS8K with a better dataset, we need to swap our eval harness. The current one is hard-coded to GSM8K.

The [perf-eval](https://github.com/vllm-project/perf-eval/tree/main) harness mentioned above supports benchmarking with [lm-eval](https://github.com/EleutherAI/lm-evaluation-harness). It has a [task named C4](https://github.com/EleutherAI/lm-evaluation-harness/tree/v0.4.12/lm_eval/tasks/c4). Is this the dataset you used for C4 news continuations, Simon? One thing missing would be a custom scorer that calls the detector to quantify detectability. Since the Detector and scorer code would live in vLLM, we might need to pass the source code with `lm_eval --include-path path/to/scorer.py`.

**Comment by simon-veitner-redhat** [comment]
@tomasruizt i think we can choose whichever dataset we like or is convenient. you should just check that detectability is given for 27B model as of now.
imo would be useful to check "how good" detector works in CI

**Comment by SyaOtiLan** [comment]
Hi @tomasruizt, I’ve implemented the optional absolute metric assertions we discussed in https://github.com/vllm-project/perf-eval/pull/80.

The current version passes 100 CPU tests. I also validated the full YAML-to-benchmark-to-assertion flow on an RTX 4090 with vLLM 0.21.0: all 224 measured requests completed, passing bounds returned success, and deliberately failing bounds retained the results and allowed subsequent configs to run before returning failure. Ingestion was tested against a loopback-only receiver, without contacting the production dashboard.

The draft includes the implementation, proposed failure policy, and validation details for review.

**Comment by tomasruizt** [comment]
@SyaOtiLan after discussing this yesterday with colleagues we determined that using the perf-eval suite (and extending its syntax with assertions) is not the simplest solution, and plain python benchmarks/tests are preferred. 

Generally a request for comment (RFC) issue is more useful before implementing anything, in particular when it's unclear what the right solution is.

**Comment by SyaOtiLan** [comment]
Understood, @tomasruizt. I moved ahead before the approach was agreed. I'll pause further work on the perf-eval extension and outline a focused Python benchmark in an RFC before implementing more. The proposal will cover watermarking on/off comparisons, measurement conditions, and the regression criterion.

RFC: https://github.com/vllm-project/vllm/issues/56968
